komposable-architecture/src/main/java/com/toggl/komposable/extensions/EffectCancellationExtensions.kt

```diff
@@ -0,0 +1,51 @@
+package com.toggl.komposable.extensions
+
+import com.toggl.komposable.architecture.Effect
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.currentCoroutineContext
+import kotlinx.coroutines.flow.cancellable
+import kotlinx.coroutines.flow.flow
+import kotlinx.coroutines.flow.onStart
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+
+private val mutex = Mutex()
+private val cancellationJobs: MutableMap<Any, MutableSet<Job>> = mutableMapOf()
+
+/**
+ * Makes an Effect cancellable by associating it with a given ID.
+ * If 'cancelInFlight' is true, any previous effects with the same ID will be cancelled.
+ *
+ * @param id The identifier for this cancellable effect. Used to manage cancellation.
+ * @param cancelInFlight If true, cancels any ongoing effects with the same ID.
+ * @return A new Effect instance that is cancellable.
+ */
+fun <Action> Effect<Action>.cancellable(id: Any, cancelInFlight: Boolean = false): Effect<Action> =
+    Effect {
+        this().onStart {
+            mutex.withLock {
+                if (cancelInFlight) {
+                    cancellationJobs[id]?.forEach { it.cancel() }
+                    cancellationJobs.remove(id)
+                }
+                val job = currentCoroutineContext()[Job]
+                job?.let { cancellationJobs.getOrPut(id) { mutableSetOf() }.add(it) }
+            }
+        }.cancellable()
+    }
+
+/**
+ * Creates an Effect that cancels any ongoing effects associated with the given action ID.
+ *
+ * @param action The ID of the effects to be cancelled.
+ * @return An Effect of Nothing that, when executed, will cancel the specified effects.
+ */
+fun Effect.Companion.cancel(action: Any): Effect<Nothing> =
```

Comment:
How is this managed in the Swift TCA? This being a singleton will make testing super hard. It's mega convenient, but a container for the cancellation ids doesn't sound like a bad idea

Reply:
If I can read Swift correctly, it's [pretty much the same thing](https://github.com/pointfreeco/swift-composable-architecture/blob/8d92762a226c486a437ef5263a0661b9112d99fa/Sources/ComposableArchitecture/Effects/Cancellation.swift#L104), right?

Reply:
Seems about correct. Let's leave it at that, then 🤷